Correctly select candidates for input solvables

lubomir opened the pull request:
If we wanted to pull in a modular package, the module should be specified. By listing a package name as input, we implicitly want a bare RPM, and even if a package with the same name is available in some module (even a default stream), we should only pull in bare rpms based on a this solvable specification.

The main use case is when Pungi expands comps group into a list of packages.

CC @contyk: is this correct? Or should default streams be allowed here as well?

igor-raits commented:
comps? how the hell is that related?

lubomir commented:
It's not visible to fus, since we expand the group to a list of packages. But it's not particularly relevant. The main point is if I ask to to include package foo, I don't think it's correct to take it from a random module that happens to contain the latest version. If a module is wanted, it should be specified explicitly.

lubomir commented:
I'll remove the reference to comps.

contyk commented:
Nack. If the package is in a default stream, you get it from that stream.

lubomir commented:
Cool, I can add a test for that, but I'm having trouble limiting the selection to only bare packages or packages from default streams.

igor-raits commented:
search for solvables which provide `module() REL_WITHOUT module-default()`

igor-raits commented:
or `REL_WITH` in your case

lubomir commented:
@r4f4 @mmathesius can one of you take a look at this? There are currently only two failing tests in this PR.

r4f4 commented:
I'll take a look at it.

r4f4 commented:
`pool_best_solvables` in `add_solvable_to_pile` will pick foo-2-1.noarch (the modular one) between the two available foo-1-1 (bare) and foo-2-1 (from non-default module).

So in `add_solvable_to_pile` should we remove packages belonging to non-default modules from the selection? When I tried subtracting from the selection, no solvables end up going to the pile:

fus-DEBUG: 12:07:28.062:  repo: fus/tests/pull-bare/packages.repo
fus-DEBUG: 12:07:28.062:  yaml: fus/tests/pull-bare/modules.yaml
fus-DEBUG: 12:07:28.062: Solvables from selection going to the pile
fus-DEBUG: 12:07:28.062:  * foo-1-1.noarch
fus-DEBUG: 12:07:28.062:  * foo-2-1.noarch
fus-DEBUG: 12:07:28.063: removing solvables from module module:platform:f29:0:00000000.x86_64
fus-DEBUG: 12:07:28.063: removing solvables from module module:m:master:20180904161631:cafebabe.x86_64
fus-DEBUG: 12:07:28.063:  * foo-2-1.noarch
fus-DEBUG: 12:07:28.063: Solvables from selection going to the pile

lubomir commented:
I was trying to do something like that, removing packages from non-default stream from the selection in `add_solvable_to_pile`. My problem was identifying the packages to remove. Can you share your code somewhere?

r4f4 committed:
Mask pkgs from non-default modules

If a solvable is provided by both a bare rpm and a non-default module,
the bare rpm is preferable even if it's an older version.

Signed-off-by: Rafael Fonseca <r4f4rfs@gmail.com>

r4f4 commented:
https://github.com/r4f4/fus/commit/59f9592747edac1e9e6137aadafbbc8731850f5b

This commit makes the first test pass.

lubomir commented:
That looks good! I think it works correctly to remove packages that are in non-default streams, which is a great improvement.

The second test needs to exclude non-modular packages if a default stream provides a package with the same name. This is very similar to #35.

lubomir commented:
I just tested it against our production data and it is sufficient to fix the problem that lead to discovering this problem.

r4f4 commented:
Maybe there is a smarter way to do it because I'm going through all modular packages when we want to go through only those matching "pkgname".

lubomir commented:
One option could be to create one big selection with all hidden packages before starting to create the pile. For each solvable it would just be subtracted and would not have to be recomputed again and again.

r4f4 commented:
While I can't come up with a better solution, I've updated the previous commit to create the selection just once.
https://github.com/r4f4/fus/commit/43c0d62ff53ea3c6994125a47fbeeadc20daac11

lubomir committed:
Add test for bare package in input

We ask to include a package by name, and it's available as a bare RPM
and in a non-default module stream. We should pull in the bare version
even though the modular one is newer.

lubomir committed:
Add test for modular package as input

We ask to include package by name. It's available as a bare RPM and in a
default module stream. The stream version should be used even though
it's older than the bare RPM.

lubomir commented:
That looks good to me. Can you open a new PR with that patch? I split the tests in this PR into two commits, so that you can include the passing test.

I think what is needed to get the other one passing is to modify the `mask_non_default_module_pkgs` function to also include bare RPMs that have the same name as packages in any default streams. But that could be done in a separate PR.

r4f4 commented:
Ok, I'll do that.

r4f4 committed:
Mask bare rpms provided by default streams

RPMs from default streams should be preferred over bare RPMs even if
they are an older version.

Signed-off-by: Rafael Fonseca <r4f4rfs@gmail.com>

lubomir commented:
With patches from @r4f4 both tests are passing. Packages from non-default streams will not be matches, and neither will bare rpms if there is a modular package with the same name available in a default stream.

lubomir committed:
Apply input masking only to selection by name

Only apply masking to packages selected by name, glob or dotarch. Exact
NEVRAs should always be included if specified.

Only one of the `selection_make` will really return something. However
this approach allows us to avoid any conditions.

lubomir commented:
I rebased this on current master. There is a new commit that only applies the filtering to packages selected by name, glob or dotarch. If the user wants explicit NEVRA, they probably really want it and it should be included. A test is added for the case as well.

r4f4 commented:
Yeah, that last commit makes sense. And the comments are spot on.

lubomir commented:
I added the matching for provides back. It's actually required to get modules in.

r4f4 commented on fus.c:
```diff
@@ -982,16 +987,49 @@ add_solvables_to_pile (Pool    *pool,
+
+      for (int i = 0; i < q.count; i++)
+        {
+          static int sel_flags = SELECTION_CANON;
```
If I had added `SELECTION_ADD` to the flags, I could have removed the Queue and `selection_add` below, simplifying it to 1 line. Not sure how to update this now though.

r4f4 replied:
I'm going to fix this in a separate commit.

r4f4 commented:
@ignatenkobrain any further comments? I plan on merging this later today.

igor-raits commented:
If it looks good for you, then go ahead.
